syntax = "proto3";
package AIBridge;

enum Role {
    DECLARER = 0;
    LOPP = 1;
    DUMMY = 2;
    ROPP = 3;
    NO = 4;
}

enum Seat {
    NORTH = 0;
    EAST = 1;
    SOUTH = 2;
    WEST = 3;
}

enum Suit {
    option allow_alias = true;
    CLUB = 0;
    C = 0;
    DIAMOND = 1;
    D = 1;
    HEART = 2;
    H = 2;
    SPADE = 3;
    S = 3;
    NOTRUMP = 4;
    NT = 4;
    DOUBLE = 5;
    X = 5;
    PASS = 6;
    P = 6;
}

enum Vul {
    NONE = 0;
    CONTRACT = 1;
    DEFENDER = 2;
    BOTH = 3;
}

message Player {
    Role role = 1;
    Seat seat = 2;
}

message Card {
    Suit suit = 1;
    uint32 rank = 2;
}

message Round {
    Player lead = 1;
    repeated Card cards = 2;
}

message CurrentSituation {
    uint32 tableID = 1;
    Vul vulnerability = 2;
    Contract contract = 3;
    Player whosTurn = 4;
    uint32 finishedDums = 5;
    uint32 contracterScore = 6;
    uint32 defenderScore = 7;
    repeated Card declarerCards = 8;
    repeated Card loppCards = 9;
    repeated Card dummyCards = 10;
    repeated Card roppCards = 11;
    repeated Round playHistory = 12;
}

message Contract {
    bool decided = 1;
    Suit suit = 2;
    uint32 level = 3;
    enum Doubled {
        NO = 0;
        X = 1;
        XX = 2;
    }
    Doubled doubled = 4;
    repeated Card contractHistory = 5;
}

message Hello {
    Player seat = 1;
    uint32 code = 2;
}

message HelloResponse {
    bool seated = 1;
    uint32 tableID = 2;
    uint32 code = 3;
}

message GameState {
    uint32 tableID = 1;
    Vul vulnerability = 2;
    Player who = 3;
    repeated Card hand = 4;
    repeated Card dummy = 5;
    repeated Round playHistory = 6;
    Contract contract = 7;
    repeated Card validPlays = 8;
}

message Play {
    uint32 tableID = 1;
    Player who = 2;
    Card card = 3;
}

message GameResult {
    uint32 tableID = 1;
    Contract contract = 2;
    int32 result = 3;
    int32 declarerScore = 4;
    int32 declarerPar = 5;
    int32 declarerIMP = 6;
}
